### ISSUE TITLE
Modernize build system to be based on pyproject.toml

### ISSUE THREAD

**Issue by anselor** [python-cmd2/cmd2#1334]
We need to replace the legacy build system based on `setup.py`, `setup.cfg` with a modern Python one based on `pyproject.toml`.  This guide on [How to modernize a setup.py based project](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/) has some useful pointers on where to start. `setup.cfg` should be deleted entirely and all configuration moved to `pyproject.toml`.  Any `python setup.py` commands need to be migrated to a modern `python -m build` or `python -m pip install` command.

There are several more advanced build system tools which are worth exploring. The current default recommendation from python seems to be [hatch](https://github.com/pypa/hatch). It has impressive features including declaring build/test configurations and managing venv environments. @anselor tried `hatch` on a couple small projects and found it to be much faster than using nox to set up venvs. The problem is it currently doesn't support a monorepo project with multiple modules or plugin modules. There is a development ticket (https://github.com/pypa/hatch/issues/233) that documents it and it is under active development now. Once that is complete we should be able to migrate to `hatch` with minimal fuss.

A newer alternative that is taking the Python world by storm is [uv](https://github.com/astral-sh/uv) which is an extremely fast Python package and project manager, written in Rust. `uv` is a single tool to replace `pip`, `pip-tools`, `pipx`, `poetry`, `pyenv`, `virtualenv`, `pipenv` and more. I've used this at work to replace `pip` in some builds and it has typically sped things up by a factor of 2x.

**Comment by tleonhardt** [comment]
I like the idea of moving to something like either [hatch](https://github.com/pypa/hatch) or [uv](https://github.com/astral-sh/uv) and getting rid of [nox](https://nox.thea.codes/en/stable/) for running our tests in multiple python environments.

I tried moving to a more modern Python project setup where everything is in `pyproject.toml` and there are no `setup.cfg` or `setup.py` files, but I couldn't get that working properly with `nox`.

**Comment by tleonhardt** [comment]
#1336 handles the moving to `ruff` from `black`, `isort`, and `flake8`

**Comment by tleonhardt** [comment]
#1343 Moved the `doc8` config to `pyrproject.toml` from `setup.cfg`

**Comment by tleonhardt** [comment]
#1344 Moves the `mypy` config to `pyproject.toml` from `setup.cfg`

**Comment by tleonhardt** [comment]
#1364 is a PR that sets the `[project]` configuration in `pyrproject.toml` and ruggedize a couple unit tests

#1365 fully removed `setup.py` and relies purely on `pyproject.toml` for the build

**Comment by tleonhardt** [comment]
I'm going to resolve this and create a new issue for considering moving to hatch or uv

**Comment by tleonhardt** [comment]
I created #1366 to encompass the future work for moving to something like `uv` or `hatch`
